apis/v1beta1/object_reference_types.go

```diff
@@ -99,9 +99,33 @@ type BackendObjectReference struct {
 	// +kubebuilder:default=""
 	Group *Group `json:"group,omitempty"`
 
-	// Kind is kind of the referent. For example "HTTPRoute" or "Service".
+	// Kind is the Kubernetes resource kind of the referent. For example
+	// "Service" or "HTTPRoute".
+	//
 	// Defaults to "Service" when not specified.
 	//
+	// Selectors in Kubernetes Services provide an abstraction around Pods,
+	// meaning that when a Service uses selectors Gateway API implementations
+	// can reason about what Namespaces the traffic will go to and can therefore
+	// enforce namespace boundaries (e.g. ReferenceGrant). As such we consider
+	// Services that utilize selectors to be considered supported when used as
+	// a BackendObjectReference at a "Core" level of support.
+	//
+	// Services which do not use Selectors (e.g. ExternalName, and headless
+	// Services without selectors) have problems with clarity about what
+	// Namespaces traffic would be destined for (and in many cases like
+	// ExternalName they can be destined in places outside the cluster
+	// entirely) and as such are considered inherently unsafe. Implementations
+	// should avoid supporting these types of Services unless they are
+	// VERY certain that they need to handle them. Due to the ambiguity
+	// described here, we consider Services which do not use Selectors to be
+	// supported under the "Implementation-specific" level of support for use as
+	// BackendObjectReferences.
+	//
+	// Support: Core (Services with Selectors)
+	//
+	// Support: Implementation-specific (Services without Selectors (e.g. ExternalName))
```

Comment:
Whether or not a Service has a Selector actually does not matter here. The presence of a selector is used to determine if Endpoint(Slice) controller(s) should create Endpoint(Slice)s for the Service. With Endpoints, leaving the selector off was the only way to attach custom self-managed endpoints. With EndpointSlices, custom self-managed endpoints can be attached to any Service, regardless of the presence of a Selector. In my opinion, the only kind of Service that should get special treatment would be ExternalName. If implementations are forwarding to ClusterIPs instead of EndpointSlices we'd also need special treatment for Headless Services, but I think I'd rather just recommend that Services forward to individual endpoints.

Reply:
> Whether or not a Service has a Selector actually does not matter here. The presence of a selector is used to determine if Endpoint(Slice) controller(s) should create Endpoint(Slice)s for the Service.

Right, this is why it _does_ matter, because it means the Endpoint(Slice)s will be made _predictably_.

> With Endpoints, leaving the selector off was the only way to attach custom self-managed endpoints. With EndpointSlices, custom self-managed endpoints can be attached to any Service, regardless of the presence of a Selector. 

Right, this is why in this draft I was considered `Selectors` the "line in the sand", on the other side of that line we can't predict what A/AAAA/CNAME entries will actually point to.

> If implementations are forwarding to ClusterIPs instead of EndpointSlices we'd also need special treatment for Headless Services, but I think I'd rather just recommend that Services forward to individual endpoints.

Wouldn't it be better to just "draw the line in the sand" and say these kinds of use cases are `Implementation-specific`, as they are hard to reason about in a _generalized manner_ due to their inherent distinctiveness?

Reply:
> Right, this is why it does matter, because it means the Endpoint(Slice)s will be made predictably.

You can have a Service with a selector and then attach user-defined EndpointSlices to it, so I don't think that is an accurate distinction.

Also EndpointSlice can also have a FQDN which behaves the same way as ExternalName. The only difference, AFAIK, is the EndpointSlice write access is considered "privileged", I think?

Reply:
You're right that it wont be true in all cases, and while I considered trying to tease this out in the language further to get the effect I was heading for, I ultimately decided that for now the better thing would be to make a meaningful language change that I expect wouldn't be controversial and gives us a bit more clarity. As such, I'm dropping this language in favor of just focusing on calling `ExternalName` type `Services` supported as `Implementation-specific` as I think that's a better baseline for now. Thanks for the thoughts and conversation on this!

Reply:
x-ref https://github.com/kubernetes/kubernetes/issues/114210 for FQDN EndpointSlices. TLDR, it's a mess. We added it without clearly defining the behavior. FQDN has been deprecated for a bit, but it may be hard to completely remove it.